plugin/pkg/global/resourcereferencemanager/admission.go

```diff
@@ -319,6 +322,97 @@ func (r *ReferenceManager) Admit(ctx context.Context, a admission.Attributes, o
 				})
 			}
 		}
+	case core.Kind("CloudProfile"):
+		cloudProfile, ok := a.GetObject().(*core.CloudProfile)
+		if !ok {
+			return apierrors.NewBadRequest("could not convert resource into CloudProfile object")
+		}
+		if utils.SkipVerification(operation, cloudProfile.ObjectMeta) {
+			return nil
+		}
+		if a.GetOperation() == admission.Update {
+			oldCloudProfile, ok := a.GetOldObject().(*core.CloudProfile)
+			if !ok {
+				return apierrors.NewBadRequest("could not convert old resource into CloudProfile object")
+			}
+
+			// getting Kubernetes versions that have been removed from the CloudProfile
+			removedKubernetesVersions := sets.StringKeySet(helper.GetRemovedVersions(oldCloudProfile.Spec.Kubernetes.Versions, cloudProfile.Spec.Kubernetes.Versions))
+
+			// getting Machine image versions that have been removed from the CloudProfile
+			removedMachineImageVersions := map[string]sets.String{}
+			for _, oldImage := range oldCloudProfile.Spec.MachineImages {
+				imageFound := false
+				for _, newImage := range cloudProfile.Spec.MachineImages {
+					if oldImage.Name == newImage.Name {
+						imageFound = true
+						removedMachineImageVersions[oldImage.Name] = sets.StringKeySet(helper.GetRemovedVersions(oldImage.Versions, newImage.Versions))
+					}
+				}
+				if !imageFound {
+					for _, version := range oldImage.Versions {
+						if removedMachineImageVersions[oldImage.Name] == nil {
+							removedMachineImageVersions[oldImage.Name] = sets.NewString()
+						}
+						removedMachineImageVersions[oldImage.Name] = removedMachineImageVersions[oldImage.Name].Insert(version.Version)
+					}
+				}
+			}
+
+			if len(removedKubernetesVersions) > 0 || len(removedMachineImageVersions) > 0 {
+				shootList, err1 := r.shootLister.List(labels.Everything())
+				if err1 != nil {
+					return apierrors.NewBadRequest(fmt.Sprintf("could not list shoots to verify that Kubernetes and/or Machine image version can be removed: %s", err1.Error()))
+				}
+
+				var (
+					channel = make(chan error)
+					wg      sync.WaitGroup
+				)
+				wg.Add(len(shootList))
+
+				for _, s := range shootList {
+					if s.Spec.CloudProfileName != cloudProfile.Name {
+						wg.Done()
+						continue
+					}
+
+					shoot := s.DeepCopy()
+					go func(shoot *core.Shoot) {
```

Comment:
I think copying the pointer should be enough here. We should avoid unnecessary calls to `DeepCopy()` in places where we loop over all Shoots (which can be a lot in some gardener installations).
```suggestion
					shoot := s
					go func() {
```

Reply:
ping